kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1218 @@
+                        Err(e)
+                    }
+                }
+            }
+            // We should never enter write for the rest of the conditions, so return a Busy error.
+            _ => Err(ErrorCode::BUSY),
+        }
+    }
+
+    fn load(&self) -> Result<(), ErrorCode> {
+        self.state.set(State::Load); // We have finished writing the last user data segment, next step is to load the process
+        match self.state.get() {
+            State::Load => {
+                match self.padding_requirement.get() {
+                    // if we decided we need to write a padding app before the new app, we go ahead and do it
+                    Some(PaddingRequirement::PrePad) | Some(PaddingRequirement::PreAndPostPad) => {
+                        // calculating the distance between our app and the previous app
+                        let previous_app_end_addr = self.previous_app_end_addr.get();
+                        let pre_pad_length = self.new_app_start_addr.get() - previous_app_end_addr;
+                        let padding_result =
+                            self.write_padding_app(pre_pad_length, previous_app_end_addr);
+                        match padding_result {
+                            Ok(()) => {
+                                if config::CONFIG.debug_load_processes {
+                                    debug!("Successfully writing prepadding app");
+                                }
+                            }
+                            Err(_e) => {
+                                // this means we were unable to write the padding app
+                                self.reset_process_loading_metadata();
+                            }
+                        };
+                    }
+                    // We should never reach here if we are not writing a prepad app
+                    None | Some(PaddingRequirement::PostPad) => {
+                        if config::CONFIG.debug_load_processes {
+                            debug!("No PrePad app to write.");
+                        }
+                    }
+                };
+                // we've written a prepad header if required, so now it is time to load the app into the
+                // process array
+                let process_flash = self.new_process_flash.take().ok_or(ErrorCode::FAIL)?;
+                let remaining_memory = self.app_memory.take().ok_or(ErrorCode::FAIL)?;
+
+                // Get the first eight bytes of flash to check if there is another app.
+                let test_header_slice = match process_flash.get(0..8) {
+                    Some(s) => s,
+                    None => {
+                        // There is no header here
+                        // if we fail here, let us erase the app we just wrote
+                        self.state.set(State::PaddingWrite);
+                        let _ = self.write_padding_app(
```

Comment:
I assume this won't work if we are already writing a padding app from line 1111?

Reply:
It works, the padding app from line 1111 only writes prepad. This line erases the app that was written and postpad by extending the prepad to the next app's start address. Essentially we rewrite the prepad such that it reclaims the memory allocated to the new app.